compute_sdk/tests/unit/test_executor.py

```diff
@@ -476,22 +515,28 @@ def test_reload_handles_deseralization_error_gracefully(gc_executor):
 
 
 @pytest.mark.parametrize("batch_size", tuple(range(1, 11)))
-def test_task_submitter_respects_batch_size(gc_executor, batch_size: int):
+def test_task_submitter_respects_batch_size(gc_executor, batch_size: int, mocker):
     gcc, gce = gc_executor
 
-    gcc.create_batch.side_effect = mock.MagicMock
-    gcc.register_function.return_value = "abc"
+    # ugly way to replace Batch with a MagicMock factory
+    mocker.patch(
+        "globus_compute_sdk.sdk.executor.Batch",
+        side_effect=lambda *_args, **_kwargs: mock.MagicMock(),
+    )
+
+    gcc.register_function.return_value = uuid.uuid4()
     num_batches = 50
 
-    gce.endpoint_id = "some_ep_id"
+    gce.endpoint_id = uuid.uuid4()
     gce.batch_size = batch_size
     for _ in range(num_batches * batch_size):
         gce.submit(noop)
-    gce.shutdown(cancel_futures=True)
+    gce.shutdown(cancel_futures=True)  # force the batches to be populated
 
+    assert gcc.batch_run.call_count >= num_batches
     for args, _kwargs in gcc.batch_run.call_args_list:
         batch, *_ = args
-        assert batch.add.call_count <= batch_size
+        assert 0 < batch.add.call_count <= batch_size
```

Comment:
This is a subtle one to understand, and shame on me for the original implementation without any supporting comments.  Given the change from `.create_batch()` to the use of imported `Batch` directly, the `side_effect=lambda` is where I land as well.  I might have written it slightly differently:

```python
mocker.patch(f"{_MOCK_BASE}Batch", side_effect=lambda *_a, **_k: mock.MagicMock())
```

but that's bikeshedding to the n<sup>th</sup> degree.

The key point for the later `assert`ions, however, is that all of the tasks have made it through the internal pipeline.  That is what the `gce.shutdown()` call effects via the default value to the `wait` parameter (`True`).

...

After extensive testing with now 3 of us, this approach turns out to be rather important due to use of `coverage`.  I'm not fully clear how coverage works, but I do know that it invokes a Python-native trace function for every statement&nbsp;&mdash;&nbsp;meaning that there's an odd timing and race-condition possibility that isn't normally there.

So ... you might consider summarizing our work over the past 24 hours into a sentence or two to make the reason for the `gce.shutdown()` less opaque/cryptic.